x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java

```diff
@@ -9,134 +9,189 @@
 import org.elasticsearch.common.hash.MessageDigests;
 import org.elasticsearch.common.settings.SecureString;
 
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
 import java.nio.charset.StandardCharsets;
 import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
 import java.util.Base64;
-import java.util.Locale;
 import java.util.Random;
 
-public enum Hasher {
+public interface Hasher {
```

Comment:
If we go with a list of predefined hash algorithms with costs, then we can set this back to an enum and have singletons that avoid extra allocations